[ROCm] CK Flash Attention Backend

alugorey opened the pull request:
Replaces https://github.com/ROCm/pytorch/pull/1592


This PR contains the initial implementation of SDPA with composable_kernel backend. The CK path can be forced by simply calling `torch.backends.cuda.preferred_rocm_fa_library("ck")`. Similarly, you can force the incumbent aotriton implementation by passing in "aotriton" or "default". As you'd expect, not setting this option will result in aotriton to be used as the backend. In the case of CK, if pytorch deems flash attention usable, then it will use the CK path in all the same places aotriton would have been used. This PR makes no changes to the heuristics which select which attention scheme to use (i.e. flash attention vs memory efficient attention vs math etc etc). It only gets called when flash attention is both enabled (via `USE_FLASH_ATTENTION`) and is selected at runtime by the existing heuristics.

Files located in pytorch/aten/src/ATen/native/transformers/hip/flash_attn/ck/mha* have been pulled from https://github.com/Dao-AILab/flash-attention courtesy of @tridao's hard work who is the co-author

NOTE: In order to use this backend, the user MUST set USE_CK_FLASH_ATTENTION=1 in their environment when they build PyTorch.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

alugorey committed:
[ROCm] CK Flash Attention Backend

Skylion007 commented on torch/backends/cuda/__init__.py:
```diff
@@ -263,9 +264,54 @@ def preferred_blas_library(
+        pass
+    elif isinstance(backend, str):
+        if backend not in _ROCmFABackends:
+            raise RuntimeError(
```
```suggestion
            raise ValueError(
```

alugorey replied:
done

drisspg commented:
Can you higlight the coverage and performance difference between this and the aottriton based version?

jithunnair-amd commented:
@alugorey This is a pretty substantive PR, esp. since it introduced a new CK backend for FA. Can you please capture the salient points in the PR description eg. changes to PyTorch build, relevant env vars if any, default settings, how new dependencies are installed, functional coverage, perf comparison vs AOTriton (if available) etc.?

pruthvistony commented on test/test_transformers.py:
```diff
@@ -3203,6 +3203,10 @@ def test_flash_attention_vs_math_ref_grads(self, device, batch_size: int, seq_le
         if max(seq_len_q, seq_len_k) >= 2048 and torch.cuda.get_device_properties('cuda').total_memory < 40 * 2**30:
             unittest.skip("Reference implementation OOM")
             return
+        if TEST_WITH_ROCM and dropout_p != 0:
```
Should we check if the backend selected is `CK`?

pruthvistony commented:
Please check the SDPA UT failure - FAILED [0.3226s] inductor/test_fused_attention.py::SDPAPatternRewriterCudaTests::test_sdpa_rewriter_10_cuda - AssertionError: wrong number of dimensions

xw285cornell commented:
This PR is very timely, thanks! Please rebase past merge conflict, add performance numbers and fix the unit test. We'd like it to get merged asap

alugorey committed:
Fix test_meta_outplace

alugorey commented:
@xw285cornell rebased the PR (as well as squashed them as all those commits became unwieldy during rebasing) and fixed the unit test errors. The fixes for those failures are in the second of the two commits. I believe I just incorrectly copied some code from a stale version of pytorch. when i updated the file to what pytorch already had all failing tests passed. I will update again once i have performance numbers.

alugorey commented:
@xw285cornell 
Here are performance numbers via https://github.com/pytorch/pytorch/blob/main/benchmarks/transformer/sdpa.py


AOTRITON:
+------------+-----------+-----------+------------+-----------+-----------+----------------+--------------------+--------------------+
| batch_size | num_heads | q_seq_len | kv_seq_len | embed_dim | is_causal |     dtype      |    forward_time    |   backward_time    |
+------------+-----------+-----------+------------+-----------+-----------+----------------+--------------------+--------------------+
|     1      |    16     |    128    |    128     |   2048    |   True    | torch.bfloat16 | 21.79451020201668  | 450.4763724980876  |
|     1      |    16     |    128    |    128     |   2048    |   False   | torch.bfloat16 | 20.076411101035774 | 417.15383948758245 |
|     1      |    16     |    256    |    256     |   2048    |   True    | torch.bfloat16 | 34.47094999719411  | 585.7575905974956  |
|     1      |    16     |    256    |    256     |   2048    |   False   | torch.bfloat16 | 25.168272096198052 | 536.6671353112906  |
|     1      |    16     |    512    |    512     |   2048    |   True    | torch.bfloat16 | 47.10616450756788  | 839.1657844185828  |
|     1      |    16     |    512    |    512     |   2048    |   False   | torch.bfloat16 | 56.71655246987939  | 800.2008404582739  |
|     1      |    16     |   1024    |    1024    |   2048    |   True    | torch.bfloat16 | 78.82721989881247  | 1465.2727136854082 |
|     1      |    16     |   1024    |    1024    |   2048    |   False   | torch.bfloat16 | 70.62346709426492  | 1246.6456997208297 |
|     8      |    16     |    128    |    128     |   2048    |   True    | torch.bfloat16 | 25.342435389757153 | 523.7751454114914  |
|     8      |    16     |    128    |    128     |   2048    |   False   | torch.bfloat16 | 21.52420529164374  | 426.74367886502296 |
|     8      |    16     |    256    |    256     |   2048    |   True    | torch.bfloat16 | 37.55110305501149  | 1003.9826692081989 |
|     8      |    16     |    256    |    256     |   2048    |   False   | torch.bfloat16 | 34.903176489751786 |  806.637704372406  |
|     8      |    16     |    512    |    512     |   2048    |   True    | torch.bfloat16 | 108.28629089519382 | 2677.0579942967743 |
|     8      |    16     |    512    |    512     |   2048    |   False   | torch.bfloat16 | 77.14254152961075  | 2057.310835225507  |
|     8      |    16     |   1024    |    1024    |   2048    |   True    | torch.bfloat16 | 340.4519024770707  | 7974.930899217727  |
|     8      |    16     |   1024    |    1024    |   2048    |   False   | torch.bfloat16 | 335.1199545431882  | 6494.933494832368  |
+------------+-----------+-----------+------------+-----------+-----------+----------------+--------------------+--------------------+

CK:
+------------+-----------+-----------+------------+-----------+-----------+----------------+--------------------+--------------------+
| batch_size | num_heads | q_seq_len | kv_seq_len | embed_dim | is_causal |     dtype      |    forward_time    |   backward_time    |
+------------+-----------+-----------+------------+-----------+-----------+----------------+--------------------+--------------------+
|     1      |    16     |    128    |    128     |   2048    |   True    | torch.bfloat16 | 17.66616189852357  | 187.1795894112438  |
|     1      |    16     |    128    |    128     |   2048    |   False   | torch.bfloat16 | 29.742619290482253 | 186.7577244993299  |
|     1      |    16     |    256    |    256     |   2048    |   True    | torch.bfloat16 | 30.34165904391557  | 186.68619252275676 |
|     1      |    16     |    256    |    256     |   2048    |   False   | torch.bfloat16 | 17.309418448712677 | 187.05361546017232 |
|     1      |    16     |    512    |    512     |   2048    |   True    | torch.bfloat16 | 31.12762535456568  | 187.47946387156844 |
|     1      |    16     |    512    |    512     |   2048    |   False   | torch.bfloat16 | 29.813363507855687 | 188.3397719357163  |
|     1      |    16     |   1024    |    1024    |   2048    |   True    | torch.bfloat16 | 57.14489193633199  | 188.11975000426173 |
|     1      |    16     |   1024    |    1024    |   2048    |   False   | torch.bfloat16 | 53.80507651716471  | 190.57486508972946 |
|     8      |    16     |    128    |    128     |   2048    |   True    | torch.bfloat16 | 30.706075858324763 | 185.7592734741047  |
|     8      |    16     |    128    |    128     |   2048    |   False   | torch.bfloat16 | 29.88582495599985  | 186.2289763521403  |
|     8      |    16     |    256    |    256     |   2048    |   True    | torch.bfloat16 | 30.88476264383644  | 188.25193995144218 |
|     8      |    16     |    256    |    256     |   2048    |   False   | torch.bfloat16 | 30.64896785654128  | 311.2350357696414  |
|     8      |    16     |    512    |    512     |   2048    |   True    | torch.bfloat16 | 67.67512508668005  |  338.631704216823  |
|     8      |    16     |    512    |    512     |   2048    |   False   | torch.bfloat16 | 72.19493587035686  | 348.0469754431396  |
|     8      |    16     |   1024    |    1024    |   2048    |   True    | torch.bfloat16 | 226.54105187393725 | 905.5793448351324  |
|     8      |    16     |   1024    |    1024    |   2048    |   False   | torch.bfloat16 | 253.85335448663685 | 925.6914001889527  |
+------------+-----------+-----------+------------+-----------+-----------+----------------+--------------------+--------------------+

Please let me know if you'd like anything else.

xw285cornell commented on aten/src/ATen/native/transformers/hip/flash_attn/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+endif()
+
+execute_process(
+  COMMAND python3 ${CMAKE_CURRENT_LIST_DIR}/../../../../../../../third_party/composable_kernel/example/ck_tile/01_fmha/generate.py
```
codegen in our internal system is kind of tricky. Any chance we can pre-generate the instance files and check them in?

alugorey replied:
technically we could, but that would require checking in over 1400+ files. Additionally, the kernels need to be regenerated every time CK is updated which would introduce some technical debt.

xw285cornell replied:
I think it's fine. This is what we do for cuda i think? https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native/transformers/cuda/flash_attn/kernels. Though 1400 is a lot :)  I'd strongly prefer checking in the file (at least for internal adoption - for glob generated files, we'll need an exact list of the files; so we'll need to update our script every single time we update CK which is a lot of work; also the CK version we use internally is different from oss pytorch.

alugorey replied:
I will check in the blob files. Are you using the latest `develop` branch for CK internally?

drisspg replied:
cc @malfet

drisspg replied:
also cc @r-barnes

xw285cornell replied:
discussed internally. Let's check in the generated files for now. We can improve in the future. But just remember we need to add those lines at the beginning

// ==========================================
// THIS CODE IS AUTOGENERATED. DO NOT MODIFY.
// @generated
// ==========================================

alugorey replied:
There are quite a bit more files than i originally quoted you. The previous number came from the first prototype. After updating CK the last time, looks like the associated generate.py script made even more than before. in any case, i committed them and added the header you suggested

jithunnair-amd replied:
@malfet Please NOTE: this PR will add 3552 kernel files to the OSS source code.

xw285cornell commented:
@alugorey it looks the backward perf is pretty bad? Do we have all the optimization in CK landed here? I thought there are some pretty good perf optimizations in CK in the past a few months.

alugorey commented:
> @alugorey thanks! it looks the backward perf is pretty bad? Do we have all the optimization in CK landed here? I thought there are some pretty good perf optimizations in CK in the past a few months.

Looks like the snapshot of CK here is from October 23rd. Quite a few changes have gone in since then. I will update CK and repost performance to see if that helps.

pruthvistony commented:
> > @alugorey thanks! it looks the backward perf is pretty bad? Do we have all the optimization in CK landed here? I thought there are some pretty good perf optimizations in CK in the past a few months.
> 
> Looks like the snapshot of CK here is from October 23rd. Quite a few changes have gone in since then. I will update CK and repost performance to see if that helps.

@xw285cornell ,
We propose to merge the current PR and take up the CK update in subsequent PR.

alugorey commented:
> > @alugorey thanks! it looks the backward perf is pretty bad? Do we have all the optimization in CK landed here? I thought there are some pretty good perf optimizations in CK in the past a few months.
> 
> Looks like the snapshot of CK here is from October 23rd. Quite a few changes have gone in since then. I will update CK and repost performance to see if that helps.

This was trickier than anticipated. Bunch of linker errors in the new kernels generated. We will follow up at a later date w.r.t. updating CK.

xw285cornell commented:
It's ok to keep this PR as is (please fix all the unit tests and populate the files - unless @drisspg disagrees :) ) But let's certainly follow up updating CK in the near future, in the end we need good perf (and internally we use a different CK version - so we cannot land this PR if it breaks with internal.

alugorey commented:
Hi @albanD , can we move around the line limit of 2000 for this PR?

malfet commented on aten/src/ATen/native/transformers/hip/flash_attn/ck/mha_fwd_ck.hip:
```diff
@@ -0,0 +1,359 @@
+/******************************************************************************
+ * Copyright (c) 2024, Tri Dao.
```
If those kernels are copy-n-paste from https://github.com/Dao-AILab/flash-attention imo it would be nice to mention in PR description,  modify respective README/LICENSE files and mention in PR that they it's co-authored by @tridao

alugorey replied:
Ah, they are from there. and you're right

alugorey committed:
Add Tri Dao to LICENSE file for functionality pulled from flash-attention

alugorey committed:
Add ck kernel blob list

alugorey committed:
Remove generated kernels from .gitignore

alugorey committed:
Add generated files to git

alugorey committed:
remove plumbing to generate files

xw285cornell commented on aten/src/ATen/native/transformers/hip/flash_attn/ck/mha_fwd_ck.hip:
```diff
@@ -0,0 +1,359 @@
+ ******************************************************************************/
+
+#include <ATen/native/transformers/hip/flash_attn/flash_common_hip.hpp>
+#include <third_party/composable_kernel/example/ck_tile/01_fmha/fmha_fwd.hpp>
```
When including ck headers, can we only use the path inside composable. Or at least get rid of the third_party?

alugorey committed:
remove examples dir from inclusion

alugorey committed:
Add files from example ck

alugorey committed:
Remove references to ck examples

xw285cornell commented:
There are a few failing tests. And for the generated files, please create a folder caffe2/aten/src/ATen/native/transformers/hip/flash_attn/instances.

jammm commented:
FWIW, this PR helped me avoid NaNs when running FA backward pass on MI300X.

alugorey committed:
Moved .h file to .hpp

xw285cornell commented:
alright I guess we cannot change to hpp because the hipify from cuda requires flash_api.h. Probably change it back, not worth it...

xw285cornell committed:
fix broken tests

alugorey committed:
Re-add file generation

alugorey committed:
Fix bad cp/paste in .gitignore

alugorey committed:
Fix lint

alugorey committed:
fix cmakelists

alugorey committed:
Remove old instance files from old location

alugorey committed:
Fix fmha_bwd/fwd.hpp

alugorey committed:
Remove old blob lists

alugorey committed:
add block_table for varlen_fwd

alugorey commented:
Converted to draft to prevent this going into release branch. Will be picked up after release branch is cut.

alugorey committed:
fix sdpUtils

jithunnair-amd commented:
@malfet Looks like the binary builds for ROCm will break due to this PR with the current changes, because CK doesn't build successfully for all gfx architectures. Our original plan was to note the increase in binary build times _on the assumption that it would build successfully_, but that assumption is turning out to be incorrect. We are currently exploring the other option we discussed, whereby we integrate a _prebuilt_ CK as a shared object into PyTorch, but that will take a few days probably.

I don't think we want to leave ROCm nightlies broken for an extended period of time while we work on the .so solution for CK, so I'm holding off merging this PR. Let me know if you have different thoughts.

cc @xw285cornell

malfet commented:
> @malfet Looks like the binary builds for ROCm will break due to this PR with the current changes, because CK doesn't build successfully for all gfx architectures. 

Can't you guard off building kernels for certain architectures, using something like `#ifdef __HIP_ARCHITECTURE == XYZ`?

jammm commented:
> @malfet Looks like the binary builds for ROCm will break due to this PR with the current changes, because CK doesn't build successfully for all gfx architectures. Our original plan was to note the increase in binary build times _on the assumption that it would build successfully_, but that assumption is turning out to be incorrect. We are currently exploring the other option we discussed, whereby we integrate a _prebuilt_ CK as a shared object into PyTorch, but that will take a few days probably.
> 
> I don't think we want to leave ROCm nightlies broken for an extended period of time while we work on the .so solution for CK, so I'm holding off merging this PR. Let me know if you have different thoughts.
> 
> cc @xw285cornell @jeffdaily @pruthvistony

FWIW, A nice middle-ground could be to force it to compile only on the supported architectures, then only switch to the CK backend if the current GPU arch is one of the supported ones.

alugorey committed:
re-Add generated instances

alugorey committed:
Remove ck cmakelists

alugorey committed:
remove calling ck cmakelists

alugorey committed:
Add blob lists

alugorey committed:
Prevent user from setting ck as backend on unsupported arch

alugorey committed:
remove generated files from .gitignore

pruthvistony commented:
@pytorchbot rebase

alugorey committed:
lint

alugorey committed:
Add guards on USE_CK_FLASH_ATTENTION

alugorey committed:
Add target_compile_definition

LunNova commented on aten/src/ATen/Context.cpp:
```diff
@@ -343,6 +343,40 @@ void Context::setBlasPreferredBackend(at::BlasBackend b) {
+  if(b == at::ROCmFABackend::Ck) {
+    static const bool ck_unsupported = []() {
+      static const std::vector<std::string> archs = {
+          "gfx90a",  "gfx942"
```
Doesn't CK support gfx908 too and gfx11xx too?

pruthvistony replied:
No it doesnt support it, currently this is the support matrix.

LunNova replied:
Ah I see. CK supports those, but the Flash Attention CK backend does not support them.

The upstream https://github.com/Dao-AILab/flash-attention library does support these gfx but it's via triton and only on with `FLASH_ATTENTION_TRITON_AMD_ENABLE`, not via the files you added here.

It's very confusing keeping track of all these support matrices :sweat_smile:

pruthvistony reviewed (dismissed):
CI is greeen.
We will work on later PR to remove the huge number of checked file into a shared library.

xw285cornell commented:
sounds good, I'll work on this internally and land it from there. Don't land from here because it's going to break

huydhn commented:
@pytorchbot merge

jammm commented:
@xw285cornell @pruthvistony is this PR available in nightly builds now?

jammm commented:
Is the nightly build built with `USE_CK_FLASH_ATTENTION=1` ?

atalman commented:
Looks like there is an issue on windows with long filename: https://github.com/pytorch/test-infra/actions/runs/12376044163/job/34542289317#step:6:447

```
  Error: error: unable to create file aten/src/ATen/native/transformers/hip/flash_attn/ck/fmha_bwd_d64_bf16_batch_b32x128x64x32x64x32x32x64x64_r1x4x1_r4x1x1_r1x4x1_w16x16x32_w16x16x16_o1_kr_ktr_vr_psddv_alibi_mask_dropout_wg16_deterministic.hip: Filename too long
```

atalman commented:
@pytorchbot revert -c weird -m "Breaks default windows checkout"

jithunnair-amd commented:
> Looks like there is an issue on windows with long filename at git checkout: https://github.com/pytorch/test-infra/actions/runs/12376044163/job/34542289317#step:6:447 when someone is using simple git checkout:
> 
> ```
>   Error: error: unable to create file aten/src/ATen/native/transformers/hip/flash_attn/ck/fmha_bwd_d64_bf16_batch_b32x128x64x32x64x32x32x64x64_r1x4x1_r4x1x1_r1x4x1_w16x16x32_w16x16x16_o1_kr_ktr_vr_psddv_alibi_mask_dropout_wg16_deterministic.hip: Filename too long
> ```
> 
> Looks like our nightlies are not affected because of the :
> 
> ```
> Set-ItemProperty -Path "HKLM:\\SYSTEM\CurrentControlSet\Control\FileSystem" -Name "LongPathsEnabled" -Value 1 step:
> 
> https://github.com/pytorch/pytorch/actions/runs/12376618338/job/34544172152?pr=143388#step:5:2
> ```
> 
> Please note: This was detected when running validations for the release because we use scripts from pytorch main however this code is not used in the release 2.6 and has no relation to release 2.6

@xw285cornell @malfet @drisspg Two suggestions to resolve this:
1. Tar these generated files and add it to pytorch/pytorch repo, and extract them only for ROCm builds when `USE_CK_FLASH_ATTENTION=1`
2. Tar these generated files and host them on repo.radeon.com and download them as part of PyTorch build only for ROCm builds when `USE_CK_FLASH_ATTENTION=1`, compare to checksum (version-controlled in PyTorch [similar to aotriton](https://github.com/pytorch/pytorch/blob/1e058a8f389b12dc8931a1bdc19c9a7ab0d73292/.ci/docker/aotriton_version.txt#L5)), extract and build

LunNova commented:
Testing this locally I'm seeing code get built for archs other than gfx90a/gfx942 for the newly added files if they're in the pytorch rocm target list.

```
[8747/9384] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/transformers/hip/flash_attn/ck/torch_hip_generated_fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi.hip.oip.o
python3.12-torch> Job completed: /build/fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi-gfx908-083330.o
python3.12-torch> Job completed: /build/fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi-gfx1100-034e9f.o
python3.12-torch> Job completed: /build/fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi-gfx90a-e70075.o
```

This is with an LLVM that has some extra logging for internal job completion which is making it obvious this is happening, by default I think you'll need to add `-v` to compiler flags to see that it built these objects and it'll be very noisy.

I'd expect not to see gfx1100 or gfx908 versions of this built; at runtime those targets aren't allowed due to the `ck_unsupported` check in Context.cpp.

jithunnair-amd committed:
Removing stale ck_kernel_blob_list.txt

jithunnair-amd committed:
Remove ck/fwd_blob_list.txt and ck/bwd_blob_list.txt because they contain local file paths and are not really necessary to be upstreamed

jithunnair-amd committed:
Rename CK autogenerated files using sha1sum to address Windows failure during git clone due to long file paths eg. https://github.com/pytorch/test-infra/actions/runs/12376044163/job/34542289317#step:6:447

jithunnair-amd committed:
Remove flash_api.h from list of includes for hipify

jithunnair-amd committed:
Add informative messages when USE_CK_FLASH_ATTENTION is enabled

jithunnair-amd committed:
Use Cmake STATUS

jithunnair-amd committed:
Add warning to users if building for more than one gfx arch in PYTORCH_ROCM_ARCH

jithunnair-amd commented:
> Testing this locally I'm seeing code get built for archs other than gfx90a/gfx942 for the newly added files if they're in the pytorch rocm target list.
> 
> ```
> [8747/9384] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/transformers/hip/flash_attn/ck/torch_hip_generated_fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi.hip.oip.o
> python3.12-torch> Job completed: /build/fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi-gfx908-083330.o
> python3.12-torch> Job completed: /build/fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi-gfx1100-034e9f.o
> python3.12-torch> Job completed: /build/fmha_fwd_d128_fp16_batch_shb_b128x128x32x128x32x128_r4x1x1_r4x1x1_w32x32x16_qr_async_vr_psskddv_alibi-gfx90a-e70075.o
> ```
> 
> This is with an LLVM that has some extra logging for internal job completion which is making it obvious this is happening, by default I think you'll need to add `-v` to compiler flags to see that it built these objects and it'll be very noisy.
> 
> I'd expect not to see gfx1100 or gfx908 versions of this built; at runtime those targets aren't allowed due to the `ck_unsupported` check in Context.cpp.

@LunNova You're right that ideally the CK targets should not even be compiled for the unsupported targets. However, that would require some build-flow changes that we wanted to keep out of the scope of this PR. We will include further build optimizations for the CK SDPA backend in subsequent PRs; this PR is intended to provide a first-level enablement of the CK SDPA backend.

Please let us know if you observe functional issues due to the above. Since trying to build the CK SDPA backend for multiple gfx archs can significantly blow up the PyTorch build time, I have included a warning message as part of this PR in [this commit](https://github.com/pytorch/pytorch/pull/138947/commits/ed20e391acab3559ec22b0c9d72e845a89462ff1).

xw285cornell commented:
Sorry, I cannot import this one due to tooling reasons, so I'm closing it and creating another one.

xw285cornell commented:
Creating another one https://github.com/pytorch/pytorch/pull/143695
